# Tarea 1 

Alumnos:

* Miguel Espinoza
* Sebastián López
* Pedro Pérez

## P1

### Parte 1

In [1]:
import os

folder = 'data/raw/'  #Carpeta principal en la que estan los archivos
weeks = os.listdir(folder)  #Se agregan las subcarpetas a una lista
weeks

['w13', 'w14', 'w15', 'w16', 'w17']

In [2]:
#Se crean dos listas para guardar los archivos normales
#y los 'furnished'
files = []
files_furnished = []

#Mediante un bucle for se guardan las direcciones de todos
#los archivos en las listas creadas
for week in weeks:
    direccion = folder + week + '/'
    file = os.listdir(direccion)
    files.append(direccion + file[0])
    files_furnished.append(direccion + file[1])
    
files

['data/raw/w13/metrocuadrado_all_w13.csv',
 'data/raw/w14/metrocuadrado_all_w14.csv',
 'data/raw/w15/metrocuadrado_all_w15.csv',
 'data/raw/w16/metrocuadrado_all_w16.csv',
 'data/raw/w17/metrocuadrado_all_w17.csv']

In [3]:
import pandas as pd

#Se lee el primer dataframe para observar los datos
data_all = pd.read_csv(files[0])
data_all.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

En primer lugar se combinan los dataframes normales

In [4]:
data_all = pd.read_csv(files[0])

for file in files:
    dato_semanal = pd.read_csv(file)
    data_all = data_all.merge(dato_semanal, how = 'outer') 
    #Se pone 'outer' para hacer la union de todos los dataframes

print('Cantidad de datos: ', len(data_all))
data_all.head(3)

Cantidad de datos:  16470


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

In [5]:
#Se ve si hay duplicados
print('Duplicados: ', data_all.duplicated().sum())

Duplicados:  175


In [6]:
#Se eliminan las filas duplicadas
data_all.drop_duplicates(inplace=True)
print('Cantidad de datos: ', len(data_all))

Cantidad de datos:  16295


Ahora se combinan los archivos furnished

In [7]:
data_furnished = pd.read_csv(files_furnished[0])

for file in files_furnished:
    dato_semanal = pd.read_csv(file)
    data_furnished = data_furnished.merge(dato_semanal, how='outer')

print('Cantidad de datos: ', len(data_furnished))
data_furnished.head(3)

Cantidad de datos:  2101


property_type|rent_type|location       price  n_rooms  \
0  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000      3.0   
1  Casa en Venta Y Arriendo, BELLAVISTA Bogotá D.C..  $1.300.000      3.0   
2  Casa en Arriendo, PORTALES DEL NORTE Bogotá D.C..  $1.800.000      3.0   

   n_bath   surface                                            details  \
0     2.0  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
1     2.0   72.0 m2  hermosa casa en venta y en arriendo en, mandal...   
2     3.0  76.81 m2  Estrato 4. CASA TRIPLEX!! Exterior . Cuenta co...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
1  https://www.metrocuadrado.com/inmueble/venta-c...                 41.4  
2  https://www.metrocuadrado.com/inmueble/arriend...                 41.3

In [8]:
#Se buscan los duplicados y se eliminan
print('Duplicados: ', data_furnished.duplicated().sum())

Duplicados:  2


In [9]:
data_furnished.drop_duplicates(inplace=True)

print('Cantidad de datos: ', len(data_furnished))
#Se observa que son muchos menos los datos de tipo furnished

Cantidad de datos:  2099


Al intentar combinar ambos dataframes usando `data_all.merge(data_furnished, how='outer')` aparece el error:

`ValueError: You are trying to merge on object and float64 columns. If you wish to proceed you should use pd.concat`

Por lo que se decide comprobar los tipos de las columnas

In [10]:
#Se observan las columnas de los dataframe
print('Tipos Data All')
print(data_all.dtypes)
print()
print('Tipos Data Furnished')
print(data_furnished.dtypes)

Tipos Data All
property_type|rent_type|location     object
price                                object
n_rooms                              object
n_bath                               object
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object

Tipos Data Furnished
property_type|rent_type|location     object
price                                object
n_rooms                             float64
n_bath                              float64
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object


Se observa que el error se debe a que en `Data All` las columnas `n_rooms` y `n_bath` son de tipo `object`, en cambio en `Data Furnished` son de tipo `float64`.

Este problema se resuelve simplemente cambiando el tipo de esas columnas.

In [11]:
#Se cambian las columnas de 'data_furnished' a object
data_furnished = data_furnished.astype({'n_rooms':'object', 'n_bath':'object'})

#Obs: No se pueden cambiar las columnas de 'data_all' a float64, pues hay un dato llamado '5+'

In [153]:
#Se combinan los dataframes
data = data_all.merge(data_furnished, how='outer', indicator='Furnished')

#Se coloca utiliza el parámetro indicator para indicar de que dataframe 
#proviene cada fila.
#left_only indica que viene del dataframe 'data_all'
#right_only indica que viene del dataframe 'data_furnished'
#y both que proviene de ambos

print('Cantidad de datos: ', len(data))
data.head(3)

Cantidad de datos:  18394


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

   Furnished  
0  left_only  
1  left_only  
2  left_only

In [154]:
data['Furnished'].replace({'left_only':'No', 'right_only':'Si', 'both':'Ambos'}, inplace=True)

#Se cambian los nombres para indicar más claramente si provienen de furnished o no
data.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished  
0        No  
1        No  
2        No

In [155]:
#Se cuentan la cantidad de cada valores de cada tipo
data['Furnished'].value_counts()

No    16295
Si     2099
Name: Furnished, dtype: int64

Como se observa no hay elementos comunes entre ambos grupos, son disjuntos. No hay archivos con texto 'furnished' que no estén contenidos en archivos con texto 'all'.

In [156]:
#Esto se comprueba tambien haciendo una intersección entre los dataframes
#mediante el argumento 'inner'
pd.merge(data_all, data_furnished, how='inner')

#Se observa que el resultado es un dataframe vacío

Empty DataFrame
Columns: [property_type|rent_type|location, price, n_rooms, n_bath, surface, details, url, metrocuadrado_index]
Index: []

### Parte 2

#### A

Primero se limpia el precio

In [157]:
data['price'].head(3)

0    $1.050.000
1    $3.930.000
2      $750.000
Name: price, dtype: object

Se observa que son de tipo `object`, pues el precio contiene el signo peso '$' y los puntos. Convendría eliminar este signo y los puntos para poder pasar los datos a `float64`.

In [158]:
data['price'] = data['price'].str.replace('$','')
data['price'] = data['price'].str.replace('.','')
data['price'].head(3)

0    1050000
1    3930000
2     750000
Name: price, dtype: object

In [159]:
#Se cambia el tipo a float64
data['price'] = data['price'].astype('float')
data['price'].head(3)

0    1050000.0
1    3930000.0
2     750000.0
Name: price, dtype: float64

Ahora se limpia el área

In [160]:
data['surface'].tail(4)

18390    164.34 m2
18391     110.0 m2
18392     106.0 m2
18393     135.0 m2
Name: surface, dtype: object

Se deben eliminar los carácteres ' m2' al final del número, y luego cambiar el tipo a `float`.

In [161]:
data['surface'] = data['surface'].str.replace(' m2','')
data['surface'] = data['surface'].astype('float')
data['surface'].tail(4)

18390    164.34
18391    110.00
18392    106.00
18393    135.00
Name: surface, dtype: float64

Ahora se limpia el número de habitaciones

In [162]:
data['n_rooms'].head(3)

0    3
1    3
2    3
Name: n_rooms, dtype: object

Parece que fueran números enteros, pero el tipo de la columna es object

In [163]:
#Se ven los valores de la columna
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, '5+', 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

Se observa que la columna tiene valores de tipo string y float. Más aún se ve que se podría cambiar inmediatamente el tipo a `float` sino existiera el valor `5+`.

In [164]:
#Se cuenta cuantas entradas hay de cada valor
data['n_rooms'].value_counts()

2      6571
1      6271
3      1351
4      1198
1.0    1040
2.0     938
5       862
3.0      52
5.0      36
4.0      32
5+        9
Name: n_rooms, dtype: int64

Se ve que solo son 9 las entradas con valor `5+`, por lo que es una parte infima del total del dataset que es de $18394$ entradas, simplemente se podrían eliminar. Pero como no se desea botar datos, se cambiará el valor `5+` por $6$, dado que es el entero mayor a 5 más cercano.

In [165]:
data['n_rooms'] = data['n_rooms'].replace('5+',6)
#Se reemplazan el valor que daba problemas, y se muestran los valores de las entradas
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, 6, 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

In [166]:
#Como esta correcto, se puede cambiar el tipo de la columna
data['n_rooms'] = data['n_rooms'].astype('float')
data['n_rooms'].head(3)

0    3.0
1    3.0
2    3.0
Name: n_rooms, dtype: float64

Por último se limpia el número de baños

In [167]:
data['n_bath'].tail(3)

18391    3
18392    3
18393    2
Name: n_bath, dtype: object

Nuevamente parecen ser enteros, pero el tipo es object

In [168]:
#Se ven los valores de la columna
data['n_bath'].unique()

array(['2', '1', '4', '3', nan, '5', '5+', 2.0, 3.0, 4.0, 5.0, 1.0],
      dtype=object)

Nuevamente el error viene dado por `5+`, por lo que se procede de igual manera que en `n_rooms`.

In [169]:
data['n_bath'] = data['n_bath'].replace('5+',6) #Se reemplazan '5+' por 6
data['n_bath'] = data['n_bath'].astype('float') #Se cambia el tipo a float
data['n_bath'].head(3)

0    2.0
1    2.0
2    1.0
Name: n_bath, dtype: float64

Finalmente se muestran nuevamente los tipos de las columnas

In [170]:
data.dtypes

property_type|rent_type|location     object
price                               float64
n_rooms                             float64
n_bath                              float64
surface                             float64
details                              object
url                                  object
metrocuadrado_index                 float64
Furnished                            object
dtype: object

#### B

Se debe dividir la columna `property_type|rent_type|location` en 3 columnas distintas.

In [171]:
#Se hace una copia de la data por si se realiza un error
copia_seguridad = data.copy()

In [444]:
#Cuando algo salga mal correr esta celda
data = copia_seguridad.copy()

In [445]:
#Se muestran unos pocos datos de esta columna
data['property_type|rent_type|location']

0          Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..
1        Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...
2        Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...
3              Casa en Arriendo, PARIS GAITAN Bogotá D.C..
4        Casa en Arriendo, BALCONES DE ORIENTE Bogotá D...
                               ...                        
18389    Apartamento en Arriendo, CHICO NORTE Bogotá D.C..
18390    Apartamento en Arriendo, CHICO NORTE II Bogotá...
18391    Apartamento en Venta Y Arriendo, CHICO NORTE B...
18392    Apartamento en Arriendo, MOLINOS NORTE Bogotá ...
18393       Apartamento en Arriendo, EL NOGAL Bogotá D.C..
Name: property_type|rent_type|location, Length: 18394, dtype: object

Se pueden deducir varias cosas:
1. El tipo de propiedad es la primera palabra del string.
2. El tipo de renta aparece después de la palabra 'en' y un espacio ' '.
3. El lugar esta después de la coma.

En primer lugar se decide dividir el string en dos partes separadas por la coma.

In [446]:
new_columns = data['property_type|rent_type|location'].str.split(pat=',', expand=True)
new_columns.head(3)

0                                           1
0  Casa en Arriendo               LA SOLEDAD NORTE Bogotá D.C..
1  Casa en Arriendo    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..
2  Casa en Arriendo   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

In [447]:
#Se insertan las columnas en la data 
data[['property_type|rent_type', 'location']] = new_columns

#Se elimina la columna original
data.drop('property_type|rent_type|location', axis=1, inplace=True)

data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type  \
0        No        Casa en Arriendo   
1        No        Casa en Arriendo   
2        No        Casa en Arriendo   

                                     location  
0               LA SOLEDAD NORTE Bogotá D.C..  
1    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..  
2   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

Se debe trabajar con la columna location para que quede sólo el nombre del barrio, es decir las palabras en mayúsculas.

In [486]:
location = data['location'].copy()
location = location.str.split()

Mediante dos bucles for se puede eliminar la ciudad de la columna location.

In [487]:
#Primero se busca en las filas
for fila in location:
    #Luego en cada palabra de la fila
    for i in range(len(fila)):
        #Si la palabra no esta mayuscula
        if not fila[i].isupper():
            #elimina la palabra y todas las palabras que le siguen de la lista
            del fila[i:]
            break

In [488]:
#Se observa el resultado
location.head()

0                   [LA, SOLEDAD, NORTE]
1        [CIUDAD, SALITRE, SUR-ORIENTAL]
2    [VILLA, DE, LOS, ALPES, II, SECTOR]
3                        [PARIS, GAITAN]
4                [BALCONES, DE, ORIENTE]
Name: location, dtype: object

Por ultimo se unen las palabras de cada lista y se añaden al dataframe.

In [491]:
location = location.str.join(' ')
data['location'] = location
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type                      location  
0        No        Casa en Arriendo              LA SOLEDAD NORTE  
1        No        Casa en Arriendo   CIUDAD SALITRE SUR-ORIENTAL  
2        No        Casa en Arriendo  VILLA DE LOS ALPES II SECTOR

Ahora se divide la columna `property_type|rent_type` en dos columnas más.

In [492]:
new_columns = data['property_type|rent_type'].str.split(pat='en ', expand=True)
new_columns.head(3)

0         1
0  Casa   Arriendo
1  Casa   Arriendo
2  Casa   Arriendo

In [493]:
#Se insertan las columnas en la data y se elimina la columna original
data[['property','rent']] = new_columns
data.drop('property_type|rent_type', axis=1, inplace=True)
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                      location property      rent  
0        No              LA SOLEDAD NORTE    Casa   Arriendo  
1        No   CIUDAD SALITRE SUR-ORIENTAL    Casa   Arriendo  
2        No  VILLA DE LOS ALPES II SECTOR    Casa   Arriendo

Por último se revisa que no hayan espacios vacíos al inicio o final de las nuevas columnas.

In [494]:
print(data['location'][0])
print(data['property'][0])
print(data['rent'][0])

LA SOLEDAD NORTE
Casa 
Arriendo


La columna property tiene espacios, por lo que se utiliza strip para eliminarlos.

In [495]:
data['property'] = data['property'].str.strip()

### Parte 3

Adjunte las siguientes columnas:
(a) Genere una variable que represente el precio por metro cuadrado.
(b) Obtenga el numero de garajes procesando la columna 'url' .

#### A

Se debe generar una nueva columna que contenga el precio por metro cuadrado.

Para esto simplemente se divide el precio por la superficie.

In [496]:
data['Precio m2'] = data['price']/data['surface']
data.head(4)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   
3  3400000.0      5.0     4.0     71.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   
3  casa ezquinera excelente ubicacion cinco alcob...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   
3  https://www.metrocuadrado.com/inmueble/arriend...                 56.0   

  Furnished                      location property      rent     Precio m2  
0        No              LA SOLEDAD NORTE     Casa  Arriendo  16666.666667  
1        No   CIUDAD SALITRE SUR-ORIENTAL     Casa  Arriendo  39300.000000  
2        No  VILLA DE LOS ALPES II SECTOR     Casa  Arriendo   8333.333333  
3        No                  PARIS GAITAN     Casa  Arriendo  47887.323944

#### B

Se debe obtener el número de garajes a partir de la columna url.

In [497]:
# Se revisa la columna url
for i in range(5):
    print(data['url'][i])

https://www.metrocuadrado.com/inmueble/arriendo-casa-bogota-la-soledad-norte-3-habitaciones-2-banos-1-garajes/35-M1919
https://www.metrocuadrado.com/inmueble/arriendo-casa-bogota-ciudad-salitre-sur-oriental-3-habitaciones-2-banos-1-garajes/769-M2387196
https://www.metrocuadrado.com/inmueble/arriendo-casa-bogota-villa-de-los-alpes-ii-sector-3-habitaciones-1-banos/BOSA-04551604
https://www.metrocuadrado.com/inmueble/arriendo-casa-bogota-paris-gaitan-5-habitaciones-4-banos/OCHE-04552802
https://www.metrocuadrado.com/inmueble/arriendo-casa-bogota-balcones-de-oriente-4-habitaciones-3-banos-1-garajes/943-M2386284


Se observa que el número de garajes aparece dentro de la url de la página.

Parece ser del estilo `n-garajes`, donde `n` es el número de garajes de la propiedad.

Sin embargo se debe tener cuidado porque hay url que no tienen garajes.

In [498]:
#En primer lugar se divide la url de acuerdo al separarador '/'
data['url'].str.split(pat='/', expand=True).head(3)

0 1                      2         3  \
0  https:    www.metrocuadrado.com  inmueble   
1  https:    www.metrocuadrado.com  inmueble   
2  https:    www.metrocuadrado.com  inmueble   

                                                   4              5  
0  arriendo-casa-bogota-la-soledad-norte-3-habita...       35-M1919  
1  arriendo-casa-bogota-ciudad-salitre-sur-orient...   769-M2387196  
2  arriendo-casa-bogota-villa-de-los-alpes-ii-sec...  BOSA-04551604

In [499]:
#Se observa que la columna 4 contiene el detalle de la vivienda
detalle = data['url'].str.split(pat='/', expand=True)[4]

#Se imprimen algunos datos de ejemplos
for i in range(5):
    print(detalle[i])

arriendo-casa-bogota-la-soledad-norte-3-habitaciones-2-banos-1-garajes
arriendo-casa-bogota-ciudad-salitre-sur-oriental-3-habitaciones-2-banos-1-garajes
arriendo-casa-bogota-villa-de-los-alpes-ii-sector-3-habitaciones-1-banos
arriendo-casa-bogota-paris-gaitan-5-habitaciones-4-banos
arriendo-casa-bogota-balcones-de-oriente-4-habitaciones-3-banos-1-garajes


Se observa que no todas las viviendas tienen garajes

In [500]:
#Se cuenta el número de filas que contiene la palabra 'garajes'
num_garajes = detalle.str.contains('garajes').sum()
print('Entradas con garajes:', num_garajes)

Entradas con garajes: 16056


Se crea una serie para guardar el numero de garajes de las viviendas

In [501]:
garajes = pd.Series(index=data.index, dtype=str, name='Garajes')
garajes

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
18389    NaN
18390    NaN
18391    NaN
18392    NaN
18393    NaN
Name: Garajes, Length: 18394, dtype: object

In [502]:
#La palabra garajes aparece al final de lista, por lo que se debe 
#obtener el penultimo valor de la lista.
#Siempre y cuando la ultima palabra de la lista sea garajes.

for i, elem in enumerate(detalle.str.split('-')):
    if elem[-1] == 'garajes':
        num = elem[-2]
    else:
        #Si no aparece, se asume que no tiene garaje
        num = 0
    garajes[i] = num

In [503]:
#Se observan los valores y el número de entradas que tiene cada uno
garajes.value_counts()

1     8091
2     6369
0     2338
3      912
4      675
4+       9
Name: Garajes, dtype: int64

Se cambia el valor `4+` por 5.

In [504]:
garajes = garajes.replace('4+',5)

Se comprueba que el número de garajes esta correcto.

In [505]:
#Al total de entradas se le restan las entradas sin garaje.
num_garajes == len(garajes) - garajes.value_counts()[0]

True

Por último, se agrega la columna a los datos.

In [506]:
data['Garajes'] = garajes
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                      location property      rent     Precio m2  \
0        No              LA SOLEDAD NORTE     Casa  Arriendo  16666.666667   
1        No   CIUDAD SALITRE SUR-ORIENTAL     Casa  Arriendo  39300.000000   
2        No  VILLA DE LOS ALPES II SECTOR     Casa  Arriendo   8333.333333   

  Garajes  
0       1  
1       1  
2       0

### Parte 4

Se deben clasificar las observaciones por tipo de producto de acuerdo a los criterios de una tabla.

In [507]:
tipo1 = data.query('property=="Casa" and surface>=80 and surface<120')
tipo2 = data.query('property=="Casa" and surface>=120 and surface<180')
tipo3 = data.query('property=="Casa" and surface>=180 and surface<240')
tipo4 = data.query('property=="Casa" and surface>=240 and surface<360')
tipo5 = data.query('property=="Casa" and surface>=360 and surface<460')
tipo6 = data.query('property=="Apartamento" and surface>=40 and surface<60')
tipo7 = data.query('property=="Apartamento" and surface>=60 and surface<80')
tipo8 = data.query('property=="Apartamento" and surface>=80 and surface<=120')

Lo que importan son los indices de estos datos.

In [509]:
t1 = pd.Series(data=1, index=tipo1.index)
t2 = pd.Series(data=2, index=tipo2.index)
t3 = pd.Series(data=3, index=tipo3.index)
t4 = pd.Series(data=4, index=tipo4.index)
t5 = pd.Series(data=5, index=tipo5.index)
t6 = pd.Series(data=6, index=tipo6.index)
t7 = pd.Series(data=7, index=tipo7.index)
t8 = pd.Series(data=8, index=tipo8.index)

Se concatenan todas estas series y se unen al dataframe.

In [510]:
cop = data.copy()

In [511]:
tipo_producto = pd.concat([t1, t2, t3, t4, t5, t6, t7, t8])
data['tipo de producto'] = tipo_producto
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                      location property      rent     Precio m2  \
0        No              LA SOLEDAD NORTE     Casa  Arriendo  16666.666667   
1        No   CIUDAD SALITRE SUR-ORIENTAL     Casa  Arriendo  39300.000000   
2        No  VILLA DE LOS ALPES II SECTOR     Casa  Arriendo   8333.333333   

  Garajes  tipo de producto  
0       1               NaN  
1       1               1.0  
2       0               1.0

Se tienen `NaN` porque hay datos que no entran en ninguno de los tipos. 

Para decidir que se hará con esos datos en primer lugar se contará el número de ellos.

In [512]:
nulos = data['tipo de producto'].isnull().sum()
print('Datos nulos:', nulos)
print('Porcentaje de nulos: {:.3f}%'.format(nulos/len(data)*100))

Datos nulos: 2774
Porcentaje de nulos: 15.081%


Como se observa es una gran cantidad de datos, por lo que se requiere tener más información para saber que hacer con esos datos.

In [513]:
data_nulos = data[data['tipo de producto'].isnull()]
data_nulos.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
3  3400000.0      5.0     4.0     71.0   
5  1100000.0      3.0     2.0     69.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
3  casa ezquinera excelente ubicacion cinco alcob...   
5  Casa en conjunto cerrado, recién remodelada, c...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
3  https://www.metrocuadrado.com/inmueble/arriend...                 56.0   
5  https://www.metrocuadrado.com/inmueble/arriend...                 32.4   

  Furnished          location property      rent     Precio m2 Garajes  \
0        No  LA SOLEDAD NORTE     Casa  Arriendo  16666.666667       1   
3        No      PARIS GAITAN     Casa  Arriendo  47887.323944       0   
5        No         TIBABUYES     Casa  Arriendo  15942.028986       0   

   tipo de producto  
0               NaN  
3               NaN  
5               NaN

Los tipos de datos que no están siendo considerados son lo siguientes:

1. Casa con área menor a $80m^2$
2. Casa con área mayor a $460m^2$
3. Apartamento con área menor a $40m^2$
4. Apartamento con área mayor a $120m^2$

Se estudiará la cantidad de datos en cada uno de estos subconjuntos.

In [514]:
tipo9 = data.query('property=="Casa" and surface<80')
tipo10 = data.query('property=="Casa" and surface>460')
tipo11 = data.query('property=="Apartamento" and surface<40')
tipo12 = data.query('property=="Apartamento" and surface>120')

In [515]:
sin_tipo = [tipo9, tipo10, tipo11, tipo12]
for i in sin_tipo:
    print('Cantidad de datos: ', len(i))

Cantidad de datos:  188
Cantidad de datos:  166
Cantidad de datos:  1100
Cantidad de datos:  1308


Se observa que la mayoria de datos que no son considerados son los de apartamentos.

In [516]:
#Se forman series con los indices de los datos sin tipo
t9 = pd.Series(data=9, index=tipo9.index)
t10 = pd.Series(data=10, index=tipo10.index)
t11 = pd.Series(data=11, index=tipo11.index)
t12 = pd.Series(data=12, index=tipo12.index)

Se concatenan todos los datos que no pertencen a ningún tipo.

In [517]:
datos_sin_tipo = pd.concat([t9, t10, t11, t12])
datos_sin_tipo.name = 'tipo de producto'

Finalmente se decide unirlos con los otros datos, pero dandoles otros tipos de acuerdo a su valor.

In [518]:
data['tipo de producto'] = data['tipo de producto'].combine_first(datos_sin_tipo)
data.head()

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   
3  3400000.0      5.0     4.0     71.0   
4  1560000.0      4.0     3.0     96.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   
3  casa ezquinera excelente ubicacion cinco alcob...   
4  Excelente casa de tres niveles, buena ubicació...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   
3  https://www.metrocuadrado.com/inmueble/arriend...                 56.0   
4  https://www.metrocuadrado.com/inmueble/arriend...                 47.1   

  Furnished                      location property      rent     Precio m2  \
0        No              LA SOLEDAD NORTE     Casa  Arriendo  16666.666667   
1        No   CIUDAD SALITRE SUR-ORIENTAL     Casa  Arriendo  39300.000000   
2        No  VILLA DE LOS ALPES II SECTOR     Casa  Arriendo   8333.333333   
3        No                  PARIS GAITAN     Casa  Arriendo  47887.323944   
4        No           BALCONES DE ORIENTE     Casa  Arriendo  16250.000000   

  Garajes  tipo de producto  
0       1               9.0  
1       1               1.0  
2       0               1.0  
3       0               9.0  
4       1               1.0

### Parte 5

A partir de la columna 'barrio', haga una fusion con el archivo data/assignacion upz/barrio-upz-
asignacion.csv para obtener as el codigo de la UPZ de cada inmueble. Reporte el numero de observaciones
y de barrios a los que no se les puede adjuntar un codigo UPZ a partir de este archivo. Tenga en cuenta
que aproximadamente 90% de los datos tiene informacion sobre UPZ.

Se cargan los datos desde `data/asignacion_upz/barrio-upz.csv`.

In [539]:
barrio = pd.read_csv('data/asignacion_upz/barrio-upz.csv')
barrio

UPlCodigo  UPlTipo          UPlNombre       UPlArea   pro_location
0      UPZ100        1           GALERIAS  2.375681e+06        palermo
1      UPZ100        1           GALERIAS  2.375681e+06     belalcazar
2      UPZ100        1           GALERIAS  2.375681e+06          sears
3      UPZ100        1           GALERIAS  2.375681e+06       galerias
4      UPZ100        1           GALERIAS  2.375681e+06  banco central
..        ...      ...                ...           ...            ...
463     UPZ42        1            VENECIA  6.638420e+06     nuevo muzu
464     UPZ42        1            VENECIA  6.638420e+06      el condor
465     UPZ42        1            VENECIA  6.638420e+06       laguneta
466     UPZ52        1           LA FLORA  1.877160e+06    villa maria
467     UPZ60        1  PARQUE ENTRENUBES  4.190395e+06            NaN

[468 rows x 5 columns]

Se deduce que la columna `UPlNombre` tiene el nombre del barrio y que se relaciona con la columna `location` del dataframe.

En primer lugar se buscan los valores únicos en la columna `UPlNombre`.

In [526]:
b_unicos = barrio['UPlNombre'].unique()
b_unicos[:4]

array(['GALERIAS', 'LAS MARGARITAS', 'QUINTA PAREDES', 'TEUSAQUILLO'],
      dtype=object)

In [654]:
code = pd.Series(data=0, index=data.index, dtype=float)

In [554]:
import numpy as np
np.nan

nan

In [584]:
location = data['location'].str.strip()

In [660]:
nombreup = barrio['UPlNombre']
codigoup = barrio['UPlCodigo']
nombreup.unique()

array(['GALERIAS', 'LAS MARGARITAS', 'QUINTA PAREDES', 'TEUSAQUILLO',
       'SAGRADO CORAZON', 'GRAN BRITALIA', 'ZONA INDUSTRIAL', 'FONTIBON',
       'DOCE DE OCTUBRE', 'MODELIA', 'PARQUE SALITRE', 'JARDIN BOTANICO',
       'EL REFUGIO', 'SUBA', 'BOLIVIA', 'COUNTRY CLUB', 'SANTA BARBARA',
       'LOS ANDES', 'SAN ISIDRO - PATIOS', 'LA ESMERALDA', 'EL PORVENIR',
       'PATIO BONITO', 'CHAPINERO', 'CIUDAD SALITRE ORIENTAL',
       'PARDO RUBIO', 'SOSIEGO', 'PUENTE ARANDA', 'LA SABANA', 'CARVAJAL',
       'LAS NIEVES', 'SANTA ISABEL', 'APOGEO', 'EL PRADO', 'NIZA',
       'CASA BLANCA SUBA', 'CIUDAD SALITRE OCCIDENTAL',
       'PARQUE SIMON BOLIVAR - CAN', 'SANTA CECILIA', 'EL TESORO',
       'MONTE BLANCO', 'TIBABUYES', 'MINUTO DE DIOS',
       'FONTIBON SAN PABLO', 'GUAYMARAL', 'CORABASTOS', 'LUCERO',
       'EL MOCHUELO', 'DIANA TURBAY', 'LOS LIBERTADORES', 'DANUBIO',
       'ARBORIZADORA', 'LA ACADEMIA', 'SAN JOSE DE BAVARIA', 'BRITALIA',
       'LAS FERIAS', 'CAPELLANIA', 'LA GLORIA

In [661]:
location.unique()
#'URBANIZACION SABANA DE TIBABUYES'

array(['LA SOLEDAD NORTE', 'CIUDAD SALITRE SUR-ORIENTAL',
       'VILLA DE LOS ALPES II SECTOR', 'PARIS GAITAN',
       'BALCONES DE ORIENTE', 'TIBABUYES', 'PORTALES DEL NORTE',
       'URBANIZACION BOLIVIA ORIENTAL', 'URBANIZACION AZAFRANES',
       'URBANIZACION EL CORTIJO', 'CAMPO ALEGRE', 'SANTA ANA ORIENTAL',
       'HIPOTECHO II', 'MIRANDELA', 'COSTA AZUL', 'CANTALEJO',
       'NORMANDIA OCCIDENTAL', 'LOS PANTANOS', 'BOSQUES DE MARIANA',
       'BOSQUE DE PINOS', 'ESTACION VICTORIA', 'VILLA ALSACIA',
       'BRITALIA', 'SAN JORGE', 'PRADO VERANIEGO NORTE',
       'URBANIZACION TIBABUYES UNIVERSAL', 'EL TINTAL A.S.D.', 'TIMIZA',
       'AUTOPISTA MUZU', 'HIPOTECHO III', 'BELLAVISTA', 'MOLINOS NORTE',
       'CEDRO BOLIVAR', 'LA LIBERIA', 'LA GRANJA', 'NUEVA ZELANDIA',
       'LA ISABELA', 'CAPELLANIA', 'VILLAS DE GRANADA I', 'BAVARIA',
       'EL REDIL', 'ACACIAS', 'BARRANCAS NORTE', 'SANTA ANA OCCIDENTAL',
       'URBANIZACION TINTALA', 'VALLADOLID', 'EL PLAN', 'SALITRE I',
     

In [662]:
location[location=='URBANIZACION SABANA DE TIBABUYES']

215      URBANIZACION SABANA DE TIBABUYES
225      URBANIZACION SABANA DE TIBABUYES
14190    URBANIZACION SABANA DE TIBABUYES
15203    URBANIZACION SABANA DE TIBABUYES
15215    URBANIZACION SABANA DE TIBABUYES
16299    URBANIZACION SABANA DE TIBABUYES
Name: location, dtype: object

In [663]:
nombreup[nombreup=='TIBABUYES']

174    TIBABUYES
175    TIBABUYES
176    TIBABUYES
Name: UPlNombre, dtype: object

In [667]:
x = location.str.contains(nombreup[174])
x[215]
print(x.argmax())
print(x[5])

5
True


In [669]:
location[215]

'URBANIZACION SABANA DE TIBABUYES'

In [675]:
codigoup[174]

'UPZ71'

In [676]:
nombreup[174]

'TIBABUYES'

In [708]:
locati = location[0:10]
locati

0                 LA SOLEDAD NORTE
1      CIUDAD SALITRE SUR-ORIENTAL
2     VILLA DE LOS ALPES II SECTOR
3                     PARIS GAITAN
4              BALCONES DE ORIENTE
5                        TIBABUYES
6                     PARIS GAITAN
7               PORTALES DEL NORTE
8    URBANIZACION BOLIVIA ORIENTAL
9           URBANIZACION AZAFRANES
Name: location, dtype: object

In [692]:
code = pd.Series(index=data.index, dtype=float)

In [710]:
for i, loc in enumerate(locati):
    for j, bar in enumerate(nombreup):
        busqueda = location.str.contains(nombreup[j])
        if busqueda.any() == True:
            indice = busqueda.argmax()
            print('i', i)
            print('loc', loc)
            print('j', j)
            print('bar', bar)
            print(nombreup[j])
            print('indice', indice)
            print(busqueda.max())
            print()
            code[indice] = codigoup[j]
            break

i 0
loc LA SOLEDAD NORTE
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 1
loc CIUDAD SALITRE SUR-ORIENTAL
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 2
loc VILLA DE LOS ALPES II SECTOR
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 3
loc PARIS GAITAN
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 4
loc BALCONES DE ORIENTE
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 5
loc TIBABUYES
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 6
loc PARIS GAITAN
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 7
loc PORTALES DEL NORTE
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 8
loc URBANIZACION BOLIVIA ORIENTAL
j 0
bar GALERIAS
GALERIAS
indice 5810
True

i 9
loc URBANIZACION AZAFRANES
j 0
bar GALERIAS
GALERIAS
indice 5810
True



In [712]:
code[code.isnull()==False]

5810    UPZ100
dtype: object

In [698]:
for i, loc in enumerate(locati):
    for j, bar in enumerate(nombreup):
        busqueda = location.str.contains(nombreup[j])
        if busqueda.any() == True:
            indice = busqueda.argmax()
            print('i', i)
            print('loc', loc)
            print('j', j)
            print('bar', bar)
            print(nombreup[j])
            print('indice', indice)
            print(busqueda.max())
            print()
            code[indice] = codigoup[j]
            break

nan

In [686]:
for i, bar in enumerate(nombreup):
    
    busqueda = location.str.contains(nombreup[i])
    if busqueda.any() == True:
        indice = busqueda.argmax()
        code[indice] = codigoup[i]

In [691]:
code.isnull().sum()

5         UPZ71
8         UPZ24
26        UPZ18
32        UPZ48
33        UPZ41
          ...  
7385      UPZ22
7564     UPZ111
9092      UPZ42
9498      UPZ81
12053     UPZ65
Length: 64, dtype: object

### Parte 6

## P2

## P3